docs/ELF_DLOPEN_METADATA.md

```diff
@@ -0,0 +1,81 @@
+---
+title: Dlopen Metadata for ELF Files
+category: Interfaces
+layout: default
+SPDX-License-Identifier: LGPL-2.1-or-later
+---
+
+# Dlopen Metadata for ELF Files
+
+*Intended audience: hackers working on packaging ELF files that use dlopen to load libraries.*
+
+## Motivation
+
+Using dlopen() to load optional dependencies brings several advantages: programs can gracefully downgrade
+when a library is not available, and the shared library is only loaded in the process (and its constructors
+are run) only when the required feature is actually used. But it also has some drawbacks, and the main one
+is that it is harder to track a program's dependencies, since unlike dynamic linking there will not be a
+mention in the ELF header. This specification aims to solve this problem by providing a standardized
+specification for a custom ELF note that can be used to list dlopen dependencies.
+
+## Implementation
+
+This document will attempt to define a common metadata format specification, so that multiple implementers
+might use it when coding upstream software, and packagers might use it when building packages and setting
+dependencies.
+
+The metadata will be embedded in a series of new, 4-bytes-aligned, allocated, 0-padded, read-only ELF header
+sections, in a name-value JSON object format, either one ELF note per dependency or as a single note using a
+json list. Implementers working on parsing ELF files should not assume a specific list of names, but parse
+anything that is included in the section, and should look for the note using the `note type`. Implementers
+working on build tools should strive to use the same names, for consistency. The most common will be listed
+here.
+
+* Section header
+
+```
+SECTION: `.note.dlopen`
+note type: `0x407c0c0a`
+Owner: `FDO` (FreeDesktop.org)
+Value: a single JSON object encoded as a zero-terminated UTF-8 string
+```
+
+* JSON payload
+
+```json
+{
+     "soname":"libfoo.so.1",
```

Comment:
The thing is that no program (at least none I was aware of) that currently implements dlopen()-based weak deps is using dlvsym(), i.e. they generally just use dlsym(), i.e. unversioned deps.

It might be worth adding a way to express explicit symbol deps, but I don't think any of the existing candidate packages would provide this info (or could easily even if it wanted). 

I can definitely say from our use in systemd we wouldn't be able to provide such info.

Reply:
I think this is not really needed. I have already implemented the tool for the Debian side, and it's easy enough and doesn't need any stubs nor symbol info. As Lennart said, we don't really use it and nobody else that I know of does either. In fact, we support multiple versions by checking if some symbols are actually available, like for libbpf.

Unlike mandatory dependencies, a suggest (which is probably the most used) will not pull in the dep anyway, so it's pointless there. And for recommends they would just be skipped if unsatisfied, so also not very useful.